Feat/pod level opa (#96)
odahu/odahu-docs: source/gen_security.rst

   protected ensures that security policy allows making this request to ODAHU API

   1. Envoy verifies JWT token in `Authorization Request Header Field`_  using `JSON Web Token (JWT) Authentication`_ filter
   2. Envoy makes a query to OpenPolicyAgent_ sidecar as PDP_ using `External Authorization`_ filter passing parsed JWT token
      from the previous step. OpenPolicyAgent_ sidecars are injected for all pods that have
      `odahu-flow-authorization=enabled` label

4. If a request is permitted by OpenPolicyAgent_, it is sent to upstream (ODAHU API)

> UML sequence diagram of successful API request described above is shown on image:

Reviewer: ```suggestion
UML sequence diagram of a successful API request described above is shown in the image:
```

.. uml:: uml/odahu_security_sequence.puml
   :scale: 90 %

***********************
Policies
***********************

ODAHU is distributed with build-in policies that are written on `Rego policy language`_ and included into helm charts
of appropriate services.

> `Role based access control`_ is implemented by default for next services

Reviewer: ```suggestion
`Role-based access control`_ is implemented by default for next services
```

   - :ref:`API <comp_api:API>`
   - :ref:`Feedback aggregator <comp_feedback:Feedback aggregator>`
   - `ODAHU deployed ML Models`

ODAHU API and Feedback aggregator policies
"""""""""""""""""""""""""""""""""""""""""""

Overview
===================

:ref:`API <comp_api:API>` and :ref:`Feedback aggregator <comp_feedback:Feedback aggregator>` are distributed
with a pre-defined set of OpenPolicyAgent_ policies. These policies implement  simple

> `Role based access control`_ (RBAC).

Reviewer: ```suggestion
`Role-based access control`_ (RBAC).
```

:ref:`installation guide <tutorials_installation:Installation>`

Customize
=========================

In this section, different ways to customize pre-defined policies

> Extend roles

Reviewer: ```suggestion
Extended roles
```

----------------------

> To define new custom roles you should just add it as a variable in file `roles.rego`

Reviewer: ```suggestion
To define new custom roles, you should add them as a variable in the `roles.rego` file
```

Create custom policies
---------------------------------

> If `Role based access control`_ is not enough for your purposes you can customize policies to
> use more general `Access based access control`_. For this rewrite `core.rego` file or create your own rego policies

Reviewer: ```suggestion
If `Role-based access control`_ is not enough for your purposes you can customize policies to
use more general `Attribute-based access control`_. For this purpose, rewrite `core.rego` file or create your own rego policies
```

   }

   allow {
       parsed_input.action == "GET"
       parsed_input.resource == "/"
   }

   allow {
       parsed_input.action == "GET"
       re_match("/swagger*", parsed_input.resource)
   }

In this file, we:

- lines 8-16: allow access if there are required permissions for action and resource for at least one user's roles
- lines 18-21: allow access to root for any user
- lines 23-26: allow access to swagger docs to any user

ODAHU ML Models pods policies
""""""""""""""""""""""""""""""""""""""

> All deployed models automatically have policies that permit requests to them to all users

Reviewer 2: ```suggestion
All deployed models contain default policies that permit requests to them to all users
```

> that have :term:`Model Deployment Access Role Name`. This role can be set while deployment of a model using

Reviewer 2: ```suggestion
that have :term:`Model Deployment Access Role Name`. This role can be set on model deployment stage using
```

> All deployed models contain default policies that permit requests to them to all users
> that have :term:`Model Deployment Access Role Name`. This role can be set on model deployment stage using
> .Spec.roleName ModelDeployment manifest key and also some other roles that statically configured in polices
> when ODAHU is deployed

Reviewer: ```suggestion
All deployed models contain default policies that permit requests to them for all users
that have :term:`Model Deployment Access Role Name`. This role can be set at the model deployment stage using
.Spec.roleName key of the ModelDeployment manifest and also statically configured in policies
during ODAHU deployment.
```
